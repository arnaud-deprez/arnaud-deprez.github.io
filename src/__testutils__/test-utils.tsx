import React from 'react'
import { render } from '@testing-library/react'
import { LocationProvider, createHistory } from '@gatsbyjs/reach-router'

const history = createHistory(window)

const AllTheProviders = ({ children }) => {
  return <LocationProvider history={history}>{children}</LocationProvider>
}

const customRender = (ui, options = {}) => render(ui, { wrapper: AllTheProviders, ...options })

// re-export everything
export * from '@testing-library/react'

// override render method
export { customRender as render }
