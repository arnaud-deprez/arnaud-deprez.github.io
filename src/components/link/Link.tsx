import React from 'react'
import { Link as GatsbyLink } from 'gatsby'
import { useLocation } from '@gatsbyjs/reach-router'
import clsx from 'clsx'

export interface InitLinkProps<T extends React.ElementType>
  extends React.AnchorHTMLAttributes<HTMLAnchorElement> {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  href: any
  locale?: string
  replace?: boolean
  activeClassName?: string
  className?: string
  role?: string
  as?: T
}

type LinkProps<T extends React.ElementType> = InitLinkProps<T> &
  Omit<React.ComponentPropsWithoutRef<T>, keyof InitLinkProps<T>>

// This assumes that any internal link (intended for Gatsby)
// will start with exactly one slash, and that anything else is external.
const isInternalRegex = /^\/(?!\/)/

const Link = React.forwardRef(function Link<T extends React.ElementType>(props: LinkProps<T>, ref) {
  const {
    activeClassName = 'active',
    className: classNameProps,
    href,
    children,
    as: As = 'a',
    ...other
  } = props

  const location = useLocation()
  const pathname = typeof href === 'string' ? href : href.pathname
  const className = clsx(classNameProps, {
    [activeClassName]: location.pathname === pathname && activeClassName,
  })

  const isInternal = isInternalRegex.test(pathname)

  if (isInternal) {
    return (
      <GatsbyLink
        className={className}
        activeClassName={activeClassName}
        to={href}
        // ref={ref}
        {...other}
      >
        {children}
      </GatsbyLink>
    )
  }

  return (
    <As className={className} href={href} ref={ref} {...other}>
      {children}
    </As>
  )
})

export default Link as <T extends React.ElementType>(props: LinkProps<T>) => JSX.Element
