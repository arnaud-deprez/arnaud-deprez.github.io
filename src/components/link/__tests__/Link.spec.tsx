import React from 'react'
import { render } from '../../../__testutils__/test-utils'
import Link from '../Link'
import { Nav } from 'react-bootstrap'

describe('<Link/>', () => {
  it('renders an internal link', () => {
    const { asFragment } = render(<Link href="/foo">Hello</Link>)
    expect(asFragment()).toMatchSnapshot()
  })
  it('renders an external link', () => {
    const { asFragment } = render(<Link href="https://foo.com/bar">Hello</Link>)
    expect(asFragment()).toMatchSnapshot()
  })
  it('renders an external link as Nav.Link', () => {
    const { asFragment } = render(
      <Link href="https://foo.com/bar" as={Nav.Link}>
        Hello
      </Link>
    )
    expect(asFragment()).toMatchSnapshot()
  })
})
