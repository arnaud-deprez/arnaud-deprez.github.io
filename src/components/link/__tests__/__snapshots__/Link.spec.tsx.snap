// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Link/> renders an external link 1`] = `
<DocumentFragment>
  <a
    class=""
    href="https://foo.com/bar"
  >
    Hello
  </a>
</DocumentFragment>
`;

exports[`<Link/> renders an external link as Nav.Link 1`] = `
<DocumentFragment>
  <a
    class="nav-link"
    href="https://foo.com/bar"
  >
    Hello
  </a>
</DocumentFragment>
`;

exports[`<Link/> renders an internal link 1`] = `
<DocumentFragment>
  <a
    class=""
    href="/foo"
  >
    Hello
  </a>
</DocumentFragment>
`;
